torch/csrc/Generator.cpp

```diff
@@ -144,9 +142,45 @@ static PyObject* THPGenerator_manualSeed(PyObject* _self, PyObject* seed) {
       throw;
     }
   }
+  // See Note [Acquire lock when using random generators]
+  std::lock_guard<std::mutex> lock(generator.mutex());
   generator.set_current_seed(seed_unpacked);
   Py_INCREF(self);
-  return (PyObject*)self;
+  return _self;
+  END_HANDLE_TH_ERRORS
+}
+
+static PyObject* THPGenerator_setOffset(PyObject* _self, PyObject* offset) {
+  HANDLE_TH_ERRORS
+  auto self = (THPGenerator*)_self;
+  auto generator = self->cdata;
+  THPUtils_assert(
+      THPUtils_checkLong(offset),
+      "manual_offset expected a long, "
+      "but got %s",
+      THPUtils_typename(offset));
+  // NOLINTNEXTLINE(cppcoreguidelines-init-variables)
+  uint64_t offset_unpacked;
+  try {
+    // First try to interpret as unsigned long
+    offset_unpacked = THPUtils_unpackUInt64(offset);
+  } catch (...) {
+    if (PyErr_ExceptionMatches(PyExc_OverflowError)) {
+      // If an overflow happened, then the offset could be negative,
+      // so try to interpret it as signed long
```

Comment:
does philox generator accept negative offsets?

Reply:
Good point. I basically copied this from looking at manual seed implementation. seed can be a negative value. 

But negative offset does not make sense to me. 

cc @albanD if you also agree, I will remove the overflow handling part.

Reply:
Well, both the seed and the offset should be `uint64_t`.
Python integer are infinite precision though so we do something weird to allow negative values.

I do agree that it doesn't matter much though, we could just pass in `n % UINT64_MAX` (the version that always returns a positive number) as well to allow any python integer value in a nicely consistent way.

Reply:
Ok, I am not changing it then. Keeping it same as `seed` to keep it consistent.